### ISSUE TITLE
Unquote support

### ISSUE THREAD

**Issue by ploeh** [hedgehogqa/fsharp-hedgehog#116]
When I write test assertions in F#, I normally prefer to use [Unquote](http://www.swensensoftware.com/unquote). I have, however, noticed that it doesn't work as well as one could wish with Hedgehog.

# Hedgehog with Unquote

Consider this simple repro:

```fsharp
open Xunit
open Hedgehog
open Swensen.Unquote

[<Fact>]
let ``Property with Unquote assertion`` () = Property.check <| property {
    let! (x, y, z) = Range.constant 0 9 |> Gen.int |> Gen.tuple3
    test <@ x + z = y + z @> // Provoke a test failure
    }
```

In this test, I've written an assertion that should cause the property to fail. I did that on purpose. When the test fails, it produces output like this:

```
Test 'HedgehogProperties.Property with Unquote assertion' failed: Swensen.Unquote.AssertionFailedException : Test failed:

3 + 3 = 7 + 3
6 = 10
false

[...]
```

I've omitted the stack trace that's also produced, because it's irrelevant for my point.

What I particularly like about Unquote is that it tells me, step-by-step, how the assertion evaluated to `false`, through a series of reductions.

That seems to work well, so what's the problem?

# Hedgehog without Unquote

The problem is that the above output doesn't show me the generated values.

Consider this alternative way to write the property, without Unquote:

```fsharp
open Xunit
open Hedgehog

[<Fact>]
let ``Property with boolean assertion`` () = Property.check <| property {
    let! (x, y, z) = Range.constant 0 9 |> Gen.int |> Gen.tuple3
    return x + z = y + z // Provoke a test failure
    }
```

When the property fails, it produces output like this:

```
Test 'HedgehogProperties.Property with boolean assertion' failed: Hedgehog.FailedException : *** Failed! Falsifiable (after 1 test and 4 shrinks):
(0, 1, 0)

[...]
```

Notice that in this case, I do get information about the values generated, the number of tests and shrinks.

# FsCheck with Unquote

FWIW, the Unquote combination works fine in FsCheck:

```fsharp
open Xunit
open FsCheck
open Swensen.Unquote

[<Fact>]
let ``Property with Unquote assertion`` () =
    let prop = Arb.generate |> Gen.three |> Arb.fromGen |> Prop.forAll <| fun (x, y, z) ->
        test <@ x + z = y + z @> // Provoke a test failure
    Check.QuickThrowOnFailure prop
```

Running this property produces output like this:

```
Test 'FsCheckProperties.Property with Unquote assertion' failed: System.Exception : Falsifiable, after 1 test (0 shrinks) (StdGen (1270052570,296354013)):
Original:
(2, 1, -2)
with exception:
Swensen.Unquote.AssertionFailedException: Test failed:

2 + -2 = 1 + -2
0 = -1
false
```

Notice that FsCheck both reports the values generated, the number of shrinks, and the Unquote assertion message.

Am I doing something wrong?

Is it possible to get the best of both worlds from Hedehog?

**Comment by moodmosaic** [comment]
Thank you for reporting this. It is a bug and can be reproduced with any property than throws an exception:
 
```f#
open Xunit
open Hedgehog

[<Fact>]
let ``Hedgehog property with effectful boolean assertion`` () = Property.check <| property {
        let! (x, y, z) = Range.constant 0 9 |> Gen.int |> Gen.tuple3
        failwith "Uh oh"
        return x + z = y + z // Provoke a test failure
        }
```
```
------ Test started: Assembly: Hedgehog.Tests.dll ------

Test 'Hedgehog.Tests.ShrinkTests.Hedgehog property with effectful boolean assertion' failed: System.Exception : Uh oh
	at Microsoft.FSharp.Core.Operators.FailWith[T](String message)
	..\fsharp-hedgehog\tests\Hedgehog.Tests\ShrinkTests.fs(19,0): at Hedgehog.Tests.ShrinkTests.Hedgehog property with effectful boolean assertion@19-1.Invoke(Tuple`3 _arg1)
	..\fsharp-hedgehog\src\Hedgehog\Property.fs(278,0): at Hedgehog.PropertyModule.prepend@278-1.Invoke(Unit _arg1)
	..\fsharp-hedgehog\src\Hedgehog\Property.fs(274,0): at Hedgehog.PropertyModule.bind@274-4.Invoke(a x)
	..\fsharp-hedgehog\src\Hedgehog\Property.fs(271,0): at Hedgehog.PropertyModule.bindGen@264.Invoke(Tuple`2 tupledArg)
	..\fsharp-hedgehog\src\Hedgehog\Gen.fs(43,0): at Hedgehog.Gen.bind@43-2.Invoke(a x)
	..\fsharp-hedgehog\src\Hedgehog\Gen.fs(40,0): at Hedgehog.Gen.bindRandom@40-1.Invoke(a x)
	..\fsharp-hedgehog\src\Hedgehog\Tree.fs(31,0): at Hedgehog.Tree.bind[a,b](Tree`1 _arg1, FSharpFunc`2 k)
	..\fsharp-hedgehog\src\Hedgehog\Gen.fs(40,0): at Hedgehog.Gen.bindRandom@33.Invoke(Seed seed0, Int32 size)
	..\fsharp-hedgehog\src\Hedgehog\Random.fs(11,0): at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1)
	..\fsharp-hedgehog\src\Hedgehog\Random.fs(14,0): at Hedgehog.Random.run[a](Seed seed, Int32 size, Random`1 r)
	..\fsharp-hedgehog\src\Hedgehog\Gen.fs(38,0): at Hedgehog.Gen.run@37T.Invoke(Seed seed, Random`1 random)
	..\fsharp-hedgehog\src\Hedgehog\Gen.fs(40,0): at Hedgehog.Gen.bindRandom@40-1.Invoke(a x)
	..\fsharp-hedgehog\src\Hedgehog\Tree.fs(31,0): at Hedgehog.Tree.bind[a,b](Tree`1 _arg1, FSharpFunc`2 k)
	..\fsharp-hedgehog\src\Hedgehog\Gen.fs(40,0): at Hedgehog.Gen.bindRandom@33.Invoke(Seed seed0, Int32 size)
	..\fsharp-hedgehog\src\Hedgehog\Random.fs(11,0): at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1)
	..\fsharp-hedgehog\src\Hedgehog\Random.fs(18,0): at Hedgehog.Random.delay@17.Invoke(Seed seed, Int32 size)
	..\fsharp-hedgehog\src\Hedgehog\Random.fs(11,0): at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1)
	..\fsharp-hedgehog\src\Hedgehog\Random.fs(14,0): at Hedgehog.Random.run[a](Seed seed, Int32 size, Random`1 r)
	..\fsharp-hedgehog\src\Hedgehog\Property.fs(320,0): at Hedgehog.PropertyModule.loop@310-20.Invoke(Seed seed, Int32 size, Int32 tests, Int32 discards)
	..\fsharp-hedgehog\src\Hedgehog\Property.fs(333,0): at Hedgehog.PropertyModule.report'(Int32 n, Property`1 p)
	..\fsharp-hedgehog\src\Hedgehog\Property.fs(336,0): at Hedgehog.PropertyModule.report(Property`1 p)
	..\fsharp-hedgehog\src\Hedgehog\Property.fs(343,0): at Hedgehog.PropertyModule.check(Property`1 p)
	..\fsharp-hedgehog\tests\Hedgehog.Tests\ShrinkTests.fs(17,0): at Hedgehog.Tests.ShrinkTests.Hedgehog property with effectful boolean assertion()

0 passed, 1 failed, 0 skipped, took 0.59 seconds (xUnit.net 2.1.0 build 3179).
```

The report should always contain the values generated, the number of shrinks, and any exception message.

That's the case with FsCheck, at least:

```f#
open Xunit
open FsCheck

[<Fact>]
let ``FsCheck property with effectful boolean assertion`` () =
    let prop = Arb.generate |> Gen.three |> Arb.fromGen |> Prop.forAll <| fun (x, y, z) ->
        failwith "Uh oh"
        x + z = y + z // Provoke a test failure
    Check.QuickThrowOnFailure prop
```

```
------ Test started: Assembly: Hedgehog.Tests.dll ------

Test 'Hedgehog.Tests.ShrinkTests.FsCheck property with effectful boolean assertion' failed: System.Exception : Falsifiable, after 1 test (0 shrinks) (StdGen (1960933309,296354613)):
Original:
(-1, -1, 2)
with exception:
System.Exception: Uh oh
   at Microsoft.FSharp.Core.Operators.FailWith[T](String message)
   at Hedgehog.Tests.ShrinkTests.prop@43.Invoke(Tuple`3 tupledArg) in ..\fsharp-hedgehog\tests\Hedgehog.Tests\ShrinkTests.fs:line 44
   at FsCheck.Testable.evaluate[a,b](FSharpFunc`2 body, a a) in C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck\Testable.fs:line 151

	at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1360.Invoke(String message)
	at <StartupCode$FsCheck>.$Runner.get_throwingRunner@346.FsCheck-IRunner-OnFinished(String , TestResult )
	at FsCheck.Runner.check[a](Config config, a p)
	at FsCheck.Check.QuickThrowOnFailure[Testable](Testable property)
	..\fsharp-hedgehog\tests\Hedgehog.Tests\ShrinkTests.fs(46,0): at Hedgehog.Tests.ShrinkTests.FsCheck property with effectful boolean assertion()

0 passed, 1 failed, 0 skipped, took 0.70 seconds (xUnit.net 2.1.0 build 3179).
```

**Comment by moodmosaic** [comment]
[FsCheck wraps the continuation passed in its `forall` in a `try...with` expression](https://github.com/fscheck/FsCheck/blob/231fc5a258d331efef8d46012b0b97120ab122af/src/FsCheck/Testable.fs#L150-L154):
![image](https://user-images.githubusercontent.com/287532/30746088-364e698c-9fb1-11e7-96c2-2c9ac77a9931.png)

This isn't the case in F# Hedgehog, which is the reason why the report is lost when an exception is thrown:
![image](https://user-images.githubusercontent.com/287532/30746174-83fa0704-9fb1-11e7-9a8e-353634dcb1d5.png)

A better way is to have `prepend` use a `try...with` expression to handle exceptions and then include their message in the report.

I haven't given it much thought yet, but this way the report should contain the values generated, the number of shrinks, and any exception message.

/cc @jystic

**Comment by moodmosaic** [comment]
@ploeh, did the fix in #119, which has been released as Hedgehog 0.4.0, worked for you?

**Comment by ploeh** [comment]
Yes, that seems to work 👍 Thank you!
